### PR TITLE
[lsp-restart]: call the force_shutdown method for the old_client

### LINKED ISSUE #3968: Lsp restart command spawn multiple procces for every call
### Summary

The new lsp restart command does not shut down the current server, so one process is left hanging, and a new one is executed.
Opening the editor for the first time: 3 processes
imagen

Running lsp-restart for the first time: 6 processes
imagen

Running lsp-restart for the second time: 9 processes
imagen

### Reproduction Steps

1. Open helix
2. Open the rust project and run the lsp-restart command once.
3. run lsp-restart a second time

### Helix log

not relevant

### Platform

Windows

### Terminal Emulator

wezterm 20220908-191626-25cd05a8

### Helix Version

helix 22.08.1 (e8f0886)

### PR DESCRIPTION
**Pull request by axdank**
fix #3968

### CODE REVIEW

File: helix-lsp/src/lib.rs [hunk 1]
```diff
@@ -356,7 +356,13 @@ impl Registry {
                 let NewClientResult(client, incoming) = start_client(id, language_config, config)?;
                 self.incoming.push(UnboundedReceiverStream::new(incoming));
 
-                entry.insert((id, client.clone()));
+                let (_, old_client) = entry.insert((id, client.clone()));
+
+                tokio::spawn(async move {
```
Comment (mangas) [inline helix-lsp/src/lib.rs:361]: I wonder if we could add this to the Drop implementation instead on the server I guess?

Comment (mangas) [reply helix-lsp/src/lib.rs:361]: I don't have a way of replicating this unfortunately, but what I'm thinking is that if you impl Drop for the client, then you could just call shutdown_and_exist, then print something if needed. This approach may also solve https://github.com/helix-editor/helix/issues/3482

File: helix-lsp/src/lib.rs [hunk 2]
```diff
@@ -356,7 +356,13 @@ impl Registry {
...
+                    if let Err(e) = old_client.shutdown_and_exit().await {
```
Comment (dead10ck) [inline helix-lsp/src/lib.rs:362]: I wonder if we should also attempt to kill the child process ourselves as well in case this fails?

**Review by dead10ck — APPROVED** [review]
One minor comment, but not a blocker, since this is a clear improvement. Thanks!

File: helix-lsp/src/lib.rs (hunk 1 above)
Comment (mangas) [reply helix-lsp/src/lib.rs:361]: I think the main issue there it's shutdown is async, I'll take a look at thsi when I have some time. This is also not a blocker, was just an improvement for better encapsulation IMO

File: helix-lsp/src/lib.rs (hunk 2 above)
Comment (dead10ck) [reply helix-lsp/src/lib.rs]: It looks like that method does not attempt to do anything with the child process, so if the LS fails to kill itself after the `shutdown` and `exit` methods, the process may linger. Maybe that would be an appropriate place to try to kill the process though.

File: helix-lsp/src/lib.rs [hunk 3]
```diff
@@ -359,7 +359,7 @@ impl Registry {
                 let (_, old_client) = entry.insert((id, client.clone()));
 
                 tokio::spawn(async move {
-                    if let Err(e) = old_client.shutdown_and_exit().await {
+                    if let Err(e) = old_client.force_shutdown().await {
```
Comment (dead10ck) [inline helix-lsp/src/lib.rs:362]: If we're opting for this instead, this method already prints an error message on shutdown failure, so we don't need one here too. It should suffice to just call the method.

**Comment by dead10ck** [comment]
I actually still think this is a good change. It shuts the LS down gracefully, using the spec's documented shutdown procedure, rather than just `kill -9`ing the process.

**Review by dead10ck — APPROVED** [review]
Looks good, thanks!

**Pull request merged**